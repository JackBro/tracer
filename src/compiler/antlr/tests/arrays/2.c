// nested arrays
main() {
	int a[10], b[5], x;

	a[2] = b[3];
	
	b[2] = a[7];

	a[3] = x + b[2];

	x = a[b[x]];

	b[a[x]] = b[a[2]];

	return a[b[x]];
}
