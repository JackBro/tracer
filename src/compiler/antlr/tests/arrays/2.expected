trans(func_main, i5,
	[p$init_0],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[p$init_1],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[],
	[],
	[],
	not_branch).
trans(func_main, i6,
	[p$init_1],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[p0],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[],
	[call(func___TRACER_INIT,[RV0|_],[SRV0|_],[])],
	[],
	not_branch).
trans(func_main, i1,
	[p0],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[p8],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A$Next$,B$Next$],[ReturnValue,X$Next$],[SA$Next$,SB$Next$],[SReturnValue,SX$Next$],
	[arrassign(SA$1$,upd(SA,2,ref(SB,3))),arrassign(SB$1$,upd(SB,2,ref(SA$1$,7))),arrassign(SA$Next$,upd(SA$1$,3,plus(SX,ref(SB$1$,2)))),assign(SX$Next$,ref(SA$Next$,ref(SB$1$,SX))),arrassign(SB$Next$,upd(SB$1$,ref(SA$Next$,SX$Next$),ref(SB$1$,ref(SA$Next$,2))))],
	[],
	[ref(A,2,ARef0),ref(B,3,BRef0),ref(B,2,BRef1),ref(A$1$,7,A$1$Ref0),ref(A$1$,3,A$1$Ref1),ref(B$1$,2,B$1$Ref0),ref(B$1$,X,B$1$Ref1),ref(A$Next$,B$1$Ref1,A$Next$Ref0),ref(B$1$,A$Next$Ref1,B$1$Ref2),ref(A$Next$,X$Next$,A$Next$Ref1),ref(A$Next$,2,A$Next$Ref2),ref(B$1$,A$Next$Ref2,B$1$Ref3)],
	not_branch) :-
	A$1$=upd(A,2,BRef0),B$1$=upd(B,2,A$1$Ref0),A$Next$=upd(A$1$,3,X+B$1$Ref0),X$Next$=A$Next$Ref0,B$Next$=upd(B$1$,A$Next$Ref1,B$1$Ref3).
trans(func_main, i2,
	[p8],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],
	[p9],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue$Next$,X],[SA,SB],[SReturnValue$Next$,SX],
	[assign(SReturnValue$Next$,ref(SA,ref(SB,SX)))],
	[],
	[ref(B,X,BRef0),ref(A,BRef0,ARef0)],
	not_branch) :-
	ReturnValue$Next$=ARef0.
trans(func___TRACER_INIT, i3,
	[p0],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[],[ReturnValue],[],[SReturnValue],
	[p1],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[],[ReturnValue],[],[SReturnValue],
	[],
	[],
	[],
	not_branch).
trans(func___TRACER_INIT, i4,
	[p1],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[],[ReturnValue],[],[SReturnValue],
	[p2],[H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[],[ReturnValue$Next$],[],[SReturnValue$Next$],
	[assign(SReturnValue$Next$,_)],
	[],
	[],
	not_branch) :-
	ReturnValue$Next$=_.
%--------------------------------------------------
block(b1, func_main, [_], [_], _, nl, [i1,i2,i5,i6],  [n,n,n,n],[],[u,u],[u,u]).
block(b2, func___TRACER_INIT, [_], [_], _, nl, [i3,i4],  [n,n,n,n],[],[],[u]).
%--------------------------------------------------
start(func_main, [p$init_0]).
start(func___TRACER_INIT, [p0]).
%--------------------------------------------------
bottom(func_main, [p9]).
bottom(func___TRACER_INIT, [p2]).
%--------------------------------------------------
freshglobalarrays([SH$1$,SH$2$,SH$3$,SH$4$]).
%--------------------------------------------------
freshglobalvars([]).
%--------------------------------------------------
freshlocalarrays(func_main,[SA,SB]).
freshlocalarrays(func___TRACER_INIT,[]).
%--------------------------------------------------
freshlocalvars(func_main,[SReturnValue,SX]).
freshlocalvars(func___TRACER_INIT,[SReturnValue]).
%--------------------------------------------------
freshparams(func_main,[lv(0)]).
freshparams(func___TRACER_INIT,[lv(0)]).
%--------------------------------------------------
init_goal(func_main, [p$init_0], i5, [H$1$,H$2$,H$3$,H$4$],[],[SH$1$,SH$2$,SH$3$,SH$4$],[],[A,B],[ReturnValue,X],[SA,SB],[SReturnValue,SX],[]).
